File: backend/app/services/tiup_service.py
import subprocess
import re
import os
import time
import logging
from typing import List, Dict, Optional
from app.models.cluster import ClusterInfo, ClusterDetail, ComponentInfo, HostInfo, LogFileInfo

logger = logging.getLogger("tiup_visualizer")

# Mapping of component role to its log file names
ROLE_LOG_FILES = {
    'tikv': ['tikv.log', 'tikv_stderr.log'],
    'pd': ['pd.log', 'pd_stderr.log'],
    'tidb': ['tidb.log', 'tidb_stderr.log'],
    'grafana': ['grafana.log'],
    'prometheus': ['prometheus.log'],
    'alertmanager': ['alertmanager.log'],
}

# Default cache TTL in seconds
CACHE_TTL = 30


class _Cache:
    """Simple TTL cache for tiup command results."""

    def __init__(self, ttl: int = CACHE_TTL):
        self._ttl = ttl
        self._store: Dict[str, tuple] = {}  # key -> (value, timestamp)

    def get(self, key: str):
        entry = self._store.get(key)
        if entry is None:
            return None
        value, ts = entry
        if time.time() - ts > self._ttl:
            del self._store[key]
            return None
        return value

    def set(self, key: str, value):
        self._store[key] = (value, time.time())

    def clear(self):
        self._store.clear()


class TiUPService:
    def __init__(self):
        self._cache = _Cache()

    @staticmethod
    def execute_command(command: str, timeout: int = 30) -> str:
        """Execute tiup command and return output"""
        try:
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=timeout
            )
            return result.stdout
        except subprocess.TimeoutExpired:
            raise Exception("Command execution timeout")
        except Exception as e:
            raise Exception(f"Command execution failed: {str(e)}")

    @staticmethod
    def get_log_files_for_role(role: str) -> List[str]:
        """Get expected log file names for a component role"""
        return ROLE_LOG_FILES.get(role.lower(), [f'{role.lower()}.log'])

    @staticmethod
    def parse_cluster_list(output: str) -> List[ClusterInfo]:
        """Parse tiup cluster list output"""
        clusters = []
        lines = output.strip().split('\n')
        
        # Skip header lines (Name, ----)
        data_lines = [line for line in lines if line and not line.startswith('Name') and not line.startswith('----')]
        
        for line in data_lines:
            parts = re.split(r'\s{2,}', line.strip())
            if len(parts) >= 5:
                clusters.append(ClusterInfo(
                    name=parts[0],
                    user=parts[1],
                    version=parts[2],
                    path=parts[3],
                    private_key=parts[4]
                ))
        
        return clusters

    @staticmethod
    def parse_cluster_display(output: str, cluster_name: str) -> ClusterDetail:
        """Parse tiup cluster display output"""
        lines = output.strip().split('\n')
        
        cluster_info = {
            'cluster_type': '',
            'cluster_name': cluster_name,
            'cluster_version': '',
            'deploy_user': '',
            'ssh_type': '',
            'dashboard_url': None,
            'grafana_url': None
        }
        
        components = []
        component_section = False
        
        for line in lines:
            line = line.strip()
            
            if not component_section:
                if line.startswith('Cluster type:'):
                    cluster_info['cluster_type'] = line.split(':', 1)[1].strip()
                elif line.startswith('Cluster version:'):
                    cluster_info['cluster_version'] = line.split(':', 1)[1].strip()
                elif line.startswith('Deploy user:'):
                    cluster_info['deploy_user'] = line.split(':', 1)[1].strip()
                elif line.startswith('SSH type:'):
                    cluster_info['ssh_type'] = line.split(':', 1)[1].strip()
                elif line.startswith('Dashboard URL:'):
                    cluster_info['dashboard_url'] = line.split(':', 1)[1].strip()
                elif line.startswith('Grafana URL:'):
                    cluster_info['grafana_url'] = line.split(':', 1)[1].strip()
                elif line.startswith('ID') and 'Role' in line:
                    component_section = True
                    continue
            else:
                if line.startswith('--') or line.startswith('Total nodes:') or not line:
                    continue
                
                parts = re.split(r'\s{2,}', line)
                if len(parts) >= 8:
                    role = parts[1]
                    log_filenames = TiUPService.get_log_files_for_role(role)
                    log_files = [LogFileInfo(filename=f, exists=True) for f in log_filenames]
                    components.append(ComponentInfo(
                        id=parts[0],
                        role=parts[1],
                        host=parts[2],
                        ports=parts[3],
                        os_arch=parts[4],
                        status=parts[5],
                        data_dir=parts[6],
                        deploy_dir=parts[7],
                        log_files=log_files
                    ))
        
        return ClusterDetail(**cluster_info, components=components)

    def get_cluster_detail(self, cluster_name: str) -> ClusterDetail:
        """Get detailed information of a specific cluster (cached)."""
        cache_key = f"detail:{cluster_name}"
        cached = self._cache.get(cache_key)
        if cached is not None:
            return cached
        output = self.execute_command(f"tiup cluster display {cluster_name}")
        detail = self.parse_cluster_display(output, cluster_name)
        self._cache.set(cache_key, detail)
        return detail

    def _get_cluster_list(self) -> List[ClusterInfo]:
        """Get raw cluster list from tiup (cached)."""
        cache_key = "cluster_list"
        cached = self._cache.get(cache_key)
        if cached is not None:
            return cached
        output = self.execute_command("tiup cluster list")
        clusters = self.parse_cluster_list(output)
        self._cache.set(cache_key, clusters)
        return clusters

    def _get_all_details(self) -> Dict[str, ClusterDetail]:
        """Fetch details for all clusters, reusing cache. Returns {name: detail}."""
        clusters = self._get_cluster_list()
        details = {}
        for cluster in clusters:
            try:
                details[cluster.name] = self.get_cluster_detail(cluster.name)
            except Exception as e:
                logger.warning(f"Error getting cluster {cluster.name} detail: {e}")
        return details

    def get_all_clusters(self) -> List[ClusterInfo]:
        """Get all tiup clusters with health status."""
        clusters = self._get_cluster_list()
        # Make copies so we don't mutate the cached list
        result = []
        details = self._get_all_details()
        for cluster in clusters:
            c = ClusterInfo(
                name=cluster.name,
                user=cluster.user,
                version=cluster.version,
                path=cluster.path,
                private_key=cluster.private_key,
            )
            detail = details.get(cluster.name)
            if detail:
                statuses = [comp.status for comp in detail.components]
                if not statuses:
                    c.status = "unknown"
                else:
                    has_up = any("Up" in s for s in statuses)
                    all_up = all("Up" in s for s in statuses)
                    if all_up:
                        c.status = "healthy"
                    elif has_up:
                        c.status = "partial"
                    else:
                        c.status = "unhealthy"
            else:
                c.status = "unknown"
            result.append(c)
        return result

    def get_all_hosts(self) -> Dict[str, HostInfo]:
        """Get all physical hosts and their components."""
        hosts_map: Dict[str, HostInfo] = {}
        details = self._get_all_details()

        for cluster_name, detail in details.items():
            for component in detail.components:
                if component.host not in hosts_map:
                    hosts_map[component.host] = HostInfo(
                        host=component.host,
                        components=[],
                        clusters=[]
                    )
                hosts_map[component.host].components.append(component)
                if cluster_name not in hosts_map[component.host].clusters:
                    hosts_map[component.host].clusters.append(cluster_name)

        return hosts_map

    def get_log_file_path(self, cluster_name: str, component_id: str, filename: str) -> str:
        """Get the local file path for a component's log file."""
        detail = self.get_cluster_detail(cluster_name)
        component = None
        for comp in detail.components:
            if comp.id == component_id:
                component = comp
                break
        if not component:
            raise Exception(f"Component {component_id} not found in cluster {cluster_name}")

        allowed_files = self.get_log_files_for_role(component.role)
        if filename not in allowed_files:
            raise Exception(f"Log file {filename} not allowed for role {component.role}")

        log_path = os.path.join(component.deploy_dir, "log", filename)
        return log_path, component

File: backend-go/tiup_service.go
package main

import (
	"context"
	"fmt"
	"log/slog"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"
)

// Mapping of component role to its expected log file names.
var roleLogFiles = map[string][]string{
	"tikv":         {"tikv.log", "tikv_stderr.log"},
	"pd":           {"pd.log", "pd_stderr.log"},
	"tidb":         {"tidb.log", "tidb_stderr.log"},
	"grafana":      {"grafana.log"},
	"prometheus":   {"prometheus.log"},
	"alertmanager": {"alertmanager.log"},
}

const cacheTTL = 30 * time.Second

// cacheEntry stores a cached value with timestamp.
type cacheEntry struct {
	value     interface{}
	timestamp time.Time
}

// ttlCache is a simple in-memory TTL cache.
type ttlCache struct {
	mu    sync.RWMutex
	store map[string]cacheEntry
	ttl   time.Duration
}

func newTTLCache(ttl time.Duration) *ttlCache {
	return &ttlCache{
		store: make(map[string]cacheEntry),
		ttl:   ttl,
	}
}

func (c *ttlCache) Get(key string) (interface{}, bool) {
	c.mu.RLock()
	entry, ok := c.store[key]
	c.mu.RUnlock()
	if !ok {
		return nil, false
	}
	if time.Since(entry.timestamp) > c.ttl {
		c.mu.Lock()
		delete(c.store, key)
		c.mu.Unlock()
		return nil, false
	}
	return entry.value, true
}

func (c *ttlCache) Set(key string, value interface{}) {
	c.mu.Lock()
	c.store[key] = cacheEntry{value: value, timestamp: time.Now()}
	c.mu.Unlock()
}

// TiUPService provides methods to interact with tiup CLI.
type TiUPService struct {
	cache *ttlCache
}

func NewTiUPService() *TiUPService {
	return &TiUPService{
		cache: newTTLCache(cacheTTL),
	}
}

// ExecuteCommand runs a shell command with timeout and returns stdout.
func ExecuteCommand(command string, timeout time.Duration) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, "bash", "-c", command)
	out, err := cmd.Output()
	if ctx.Err() == context.DeadlineExceeded {
		return "", fmt.Errorf("command execution timeout")
	}
	if err != nil {
		return "", fmt.Errorf("command execution failed: %w", err)
	}
	return string(out), nil
}

// GetLogFilesForRole returns expected log file names for a component role.
func GetLogFilesForRole(role string) []string {
	r := strings.ToLower(role)
	if files, ok := roleLogFiles[r]; ok {
		return files
	}
	return []string{r + ".log"}
}

var splitRegex = regexp.MustCompile(`\s{2,}`)

// ParseClusterList parses the output of `tiup cluster list`.
func ParseClusterList(output string) []ClusterInfo {
	var clusters []ClusterInfo
	lines := strings.Split(strings.TrimSpace(output), "\n")

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "Name") || strings.HasPrefix(line, "----") {
			continue
		}
		parts := splitRegex.Split(line, -1)
		if len(parts) >= 5 {
			clusters = append(clusters, ClusterInfo{
				Name:       parts[0],
				User:       parts[1],
				Version:    parts[2],
				Path:       parts[3],
				PrivateKey: parts[4],
				Status:     "unknown",
			})
		}
	}
	return clusters
}

// ParseClusterDisplay parses the output of `tiup cluster display <name>`.
func ParseClusterDisplay(output string, clusterName string) *ClusterDetail {
	lines := strings.Split(strings.TrimSpace(output), "\n")

	detail := &ClusterDetail{
		ClusterName: clusterName,
		Components:  []ComponentInfo{},
	}

	componentSection := false
	for _, line := range lines {
		line = strings.TrimSpace(line)

		if !componentSection {
			if strings.HasPrefix(line, "Cluster type:") {
				detail.ClusterType = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
			} else if strings.HasPrefix(line, "Cluster version:") {
				detail.ClusterVersion = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
			} else if strings.HasPrefix(line, "Deploy user:") {
				detail.DeployUser = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
			} else if strings.HasPrefix(line, "SSH type:") {
				detail.SSHType = strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
			} else if strings.HasPrefix(line, "Dashboard URL:") {
				v := strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
				detail.DashboardURL = &v
			} else if strings.HasPrefix(line, "Grafana URL:") {
				v := strings.TrimSpace(strings.SplitN(line, ":", 2)[1])
				detail.GrafanaURL = &v
			} else if strings.HasPrefix(line, "ID") && strings.Contains(line, "Role") {
				componentSection = true
				continue
			}
		} else {
			if strings.HasPrefix(line, "--") || strings.HasPrefix(line, "Total nodes:") || line == "" {
				continue
			}
			parts := splitRegex.Split(line, -1)
			if len(parts) >= 8 {
				role := parts[1]
				logFilenames := GetLogFilesForRole(role)
				logFiles := make([]LogFileInfo, len(logFilenames))
				for i, f := range logFilenames {
					logFiles[i] = LogFileInfo{Filename: f, Exists: true}
				}
				detail.Components = append(detail.Components, ComponentInfo{
					ID:        parts[0],
					Role:      parts[1],
					Host:      parts[2],
					Ports:     parts[3],
					OSArch:    parts[4],
					Status:    parts[5],
					DataDir:   parts[6],
					DeployDir: parts[7],
					LogFiles:  logFiles,
				})
			}
		}
	}
	return detail
}

func (s *TiUPService) GetClusterDetail(clusterName string) (*ClusterDetail, error) {
	cacheKey := "detail:" + clusterName
	if v, ok := s.cache.Get(cacheKey); ok {
		return v.(*ClusterDetail), nil
	}

	output, err := ExecuteCommand(fmt.Sprintf("tiup cluster display %s", clusterName), 30*time.Second)
	if err != nil {
		return nil, err
	}
	detail := ParseClusterDisplay(output, clusterName)
	s.cache.Set(cacheKey, detail)
	return detail, nil
}

func (s *TiUPService) getClusterList() ([]ClusterInfo, error) {
	cacheKey := "cluster_list"
	if v, ok := s.cache.Get(cacheKey); ok {
		return v.([]ClusterInfo), nil
	}

	output, err := ExecuteCommand("tiup cluster list", 30*time.Second)
	if err != nil {
		return nil, err
	}
	clusters := ParseClusterList(output)
	s.cache.Set(cacheKey, clusters)
	return clusters, nil
}

func (s *TiUPService) getAllDetails() map[string]*ClusterDetail {
	clusters, err := s.getClusterList()
	if err != nil {
		slog.Warn("Error getting cluster list", "error", err)
		return nil
	}
	details := make(map[string]*ClusterDetail)
	for _, c := range clusters {
		detail, err := s.GetClusterDetail(c.Name)
		if err != nil {
			slog.Warn("Error getting cluster detail", "cluster", c.Name, "error", err)
			continue
		}
		details[c.Name] = detail
	}
	return details
}

// GetAllClusters returns all clusters with computed health status.
func (s *TiUPService) GetAllClusters() ([]ClusterInfo, error) {
	clusters, err := s.getClusterList()
	if err != nil {
		return nil, err
	}

	details := s.getAllDetails()
	result := make([]ClusterInfo, len(clusters))
	for i, c := range clusters {
		result[i] = ClusterInfo{
			Name:       c.Name,
			User:       c.User,
			Version:    c.Version,
			Path:       c.Path,
			PrivateKey: c.PrivateKey,
			Status:     "unknown",
		}

		if detail, ok := details[c.Name]; ok && len(detail.Components) > 0 {
			hasUp := false
			allUp := true
			for _, comp := range detail.Components {
				if strings.Contains(comp.Status, "Up") {
					hasUp = true
				} else {
					allUp = false
				}
			}
			if allUp {
				result[i].Status = "healthy"
			} else if hasUp {
				result[i].Status = "partial"
			} else {
				result[i].Status = "unhealthy"
			}
		}
	}
	return result, nil
}

// GetAllHosts aggregates components by physical host.
func (s *TiUPService) GetAllHosts() (map[string]*HostInfo, error) {
	details := s.getAllDetails()
	hostsMap := make(map[string]*HostInfo)

	for clusterName, detail := range details {
		for _, comp := range detail.Components {
			host, ok := hostsMap[comp.Host]
			if !ok {
				host = &HostInfo{
					Host:       comp.Host,
					Components: []ComponentInfo{},
					Clusters:   []string{},
				}
				hostsMap[comp.Host] = host
			}
			host.Components = append(host.Components, comp)
			// Add cluster name if not already present
			found := false
			for _, cn := range host.Clusters {
				if cn == clusterName {
					found = true
					break
				}
			}
			if !found {
				host.Clusters = append(host.Clusters, clusterName)
			}
		}
	}
	return hostsMap, nil
}

// GetLogFilePath returns the log file path and the component info.
func (s *TiUPService) GetLogFilePath(clusterName, componentID, filename string) (string, *ComponentInfo, error) {
	detail, err := s.GetClusterDetail(clusterName)
	if err != nil {
		return "", nil, err
	}

	var component *ComponentInfo
	for i := range detail.Components {
		if detail.Components[i].ID == componentID {
			component = &detail.Components[i]
			break
		}
	}
	if component == nil {
		return "", nil, fmt.Errorf("component %s not found in cluster %s", componentID, clusterName)
	}

	allowedFiles := GetLogFilesForRole(component.Role)
	allowed := false
	for _, f := range allowedFiles {
		if f == filename {
			allowed = true
			break
		}
	}
	if !allowed {
		return "", nil, fmt.Errorf("log file %s not allowed for role %s", filename, component.Role)
	}

	logPath := filepath.Join(component.DeployDir, "log", filename)
	return logPath, component, nil
}
